File: api/routers/metrics.py
from fastapi import APIRouter, HTTPException

from api.database import fetch_all
from api.schemas import (
    EcommerceOverview,
    EventActivity,
    ConversionFunnel,
    PaymentPerformance,
    ProductPerformance,
)

router = APIRouter(
    prefix="/api/metrics",
    tags=["metrics"],
)


# --------------------------------------------------
# 1. Ecommerce Overview
# --------------------------------------------------


@router.get(
    "/overview",
    response_model=EcommerceOverview,
)
def get_overview():

    query = """
        SELECT *
        FROM workspace.gold.ecommerce_overview
    """

    rows = fetch_all(query)

    if not rows:
        raise HTTPException(
            status_code=404,
            detail="No overview metrics available",
        )

    return rows[0]


# --------------------------------------------------
# 2. Product Performance
# --------------------------------------------------


@router.get(
    "/products",
    response_model=list[ProductPerformance],
)
def get_product_performance():

    query = """
        SELECT *
        FROM workspace.gold.product_performance
        ORDER BY successful_revenue DESC
    """

    rows = fetch_all(query)

    if not rows:
        raise HTTPException(
            status_code=404,
            detail="No product performance metrics available",
        )

    return rows


# --------------------------------------------------
# 3. Payment Performance
# --------------------------------------------------


@router.get(
    "/payments",
    response_model=list[PaymentPerformance],
)
def get_payment_performance():

    query = """
        SELECT *
        FROM workspace.gold.payment_performance
        ORDER BY total_attempts DESC
    """

    rows = fetch_all(query)

    if not rows:
        raise HTTPException(
            status_code=404,
            detail="No payment performance metrics available",
        )

    return rows


# --------------------------------------------------
# 4. Conversion Funnel
# --------------------------------------------------


@router.get(
    "/conversion",
    response_model=ConversionFunnel,
)
def get_conversion_funnel():

    query = """
        SELECT *
        FROM workspace.gold.conversion_funnel
    """

    rows = fetch_all(query)

    if not rows:
        raise HTTPException(
            status_code=404,
            detail="No conversion funnel metrics available",
        )

    return rows[0]


# --------------------------------------------------
# 5. Event Activity
# --------------------------------------------------


@router.get(
    "/activity",
    response_model=list[EventActivity],
)
def get_event_activity():

    query = """
        SELECT *
        FROM workspace.gold.event_activity
        ORDER BY window_start, event_type
    """

    rows = fetch_all(query)

    if not rows:
        raise HTTPException(
            status_code=404,
            detail="No event activity metrics available",
        )

    return rows

File: dashboard/lib/api/metrics.ts
const METRICS_API_URL =
  process.env.NEXT_PUBLIC_METRICS_API_URL ?? "http://127.0.0.1:8001";

async function fetchMetrics<T>(endpoint: string): Promise<T> {
  const response = await fetch(`${METRICS_API_URL}${endpoint}`);

  if (!response.ok) {
    let message = `Metrics API request failed: ${response.status}`;

    try {
      const body = await response.json();

      if (typeof body.detail === "string") {
        message = body.detail;
      }
    } catch {
      // Keep the default error message.
    }

    throw new Error(message);
  }

  return response.json();
}

// --------------------------------------------------
// Ecommerce Overview
// --------------------------------------------------

export interface EcommerceOverview {
  total_users: number;
  total_product_views: number;
  unique_viewers: number;
  total_orders: number;
  total_payment_attempts: number;
  successful_payments: number;
  order_value: number;
  successful_revenue: number;
  average_order_value: number;
  view_to_order_rate: number;
  payment_success_rate: number;
}

export function getOverview(): Promise<EcommerceOverview> {
  return fetchMetrics<EcommerceOverview>("/api/metrics/overview");
}

// --------------------------------------------------
// Product Performance
// --------------------------------------------------

export interface ProductPerformance {
  product_id: string;
  product_name: string;
  category: string;
  views: number;
  unique_viewers: number;
  orders: number;
  units_sold: number;
  order_value: number;
  successful_revenue: number;
  conversion_rate: number;
}

export function getProductPerformance(): Promise<ProductPerformance[]> {
  return fetchMetrics<ProductPerformance[]>("/api/metrics/products");
}

// --------------------------------------------------
// Payment Performance
// --------------------------------------------------

export interface PaymentPerformance {
  payment_mode: string;
  payment_provider: string;
  total_attempts: number;
  successful_attempts: number;
  failed_attempts: number;
  pending_attempts: number;
  total_payment_amount: number;
  successful_payment_amount: number;
  success_rate: number;
  failure_rate: number;
  pending_rate: number;
}

export function getPaymentPerformance(): Promise<PaymentPerformance[]> {
  return fetchMetrics<PaymentPerformance[]>("/api/metrics/payments");
}

// --------------------------------------------------
// Conversion Funnel
// --------------------------------------------------

export interface ConversionFunnel {
  total_users: number;
  total_product_views: number;
  total_orders: number;
  total_payment_attempts: number;
  successful_payments: number;
  views_per_user: number;
  view_to_order_rate: number;
  payment_attempts_per_order: number;
  payment_success_rate: number;
  view_to_successful_payment_rate: number;
}

export function getConversionFunnel(): Promise<ConversionFunnel> {
  return fetchMetrics<ConversionFunnel>("/api/metrics/conversion");
}

// --------------------------------------------------
// Event Activity
// --------------------------------------------------

export interface EventActivity {
  window_start: string;
  window_end: string;
  event_type: string;
  event_count: number;
  unique_users: number;
  events_per_active_user: number;
}

export function getEventActivity(): Promise<EventActivity[]> {
  return fetchMetrics<EventActivity[]>("/api/metrics/activity");
}
